common/components/org.wso2.carbon.data.agents.log4j/src/main/java/org/wso2/carbon/data/agents/log4j/appender/LogEventAppender.java

```diff
@@ -0,0 +1,411 @@
+                            getMaxTolerableConsecutiveFailure() + ". Purging log event array. Some logs will be lost.");
+                    loggingEvents.clear();
+                    numOfConsecutiveFailures = 0;
+                }
+            }
+        }
+
+        private void publishLogEvent(TenantAwareLoggingEvent event)
+                throws ParseException, MalformedURLException, AuthenticationException, TransportException,
+                MalformedStreamDefinitionException, StreamDefinitionException,
+                DifferentStreamDefinitionAlreadyDefinedException, ExecutionException {
+            String streamId = "";
+            String streamName = "";
+            String tenantId = event.getTenantId();
+            if (tenantId.equals(String.valueOf(MultitenantConstants.INVALID_TENANT_ID)) || tenantId
+                    .equals(String.valueOf(MultitenantConstants.SUPER_TENANT_ID))) {
+                tenantId = "0";
+            }
+            String serverKey = getCurrentServerName();
+            String currDateStr = getCurrentDate();
+            try {
+                Pattern pattern = Pattern.compile("(tcp):\\/\\/([a-zA-Z0-9]+):([0-9]+)");
+                Matcher matcher = pattern.matcher(url);
+                if (matcher.find() && authURLSet == null) {
+                    authURLSet = "ssl://" + matcher.group(2).toString() + ":" + (
+                            Integer.parseInt(matcher.group(3).toString()) + 100);
+                }
+                dataPublisher = new DataPublisher("Thrift", url, authURLSet, userName, password);
+            } catch (DataEndpointAgentConfigurationException e) {
+                log.error(
+                        "Error in initializing binary data-publisher to send requests to global throttling engine " + e
+                                .getMessage(), e);
+            } catch (DataEndpointException e) {
+                log.error(
+                        "Error in initializing binary data-publisher to send requests to global throttling engine " + e
+                                .getMessage(), e);
+            } catch (DataEndpointConfigurationException e) {
+                log.error(
+                        "Error in initializing binary data-publisher to send requests to global throttling engine " + e
+                                .getMessage(), e);
+            } catch (DataEndpointAuthenticationException e) {
+                log.error(
+                        "Error in initializing binary data-publisher to send requests to global throttling engine " + e
+                                .getMessage(), e);
+            } catch (TransportException e) {
+                log.error(
+                        "Error in initializing binary data-publisher to send requests to global throttling engine " + e
+                                .getMessage(), e);
+            }
```

Comment:
Hi Krishantha,

Thank you for feedback regarding the issue when data publisher not properly initialize,

When the publisher not initialize it will return NPE. Then it will catch by the Throwable clause in https://github.com/wso2/shared-analytics/pull/126/files#diff-d45f010b040aded5b5711f7ddab28e59R385 and it will clean loggingEvents queue. 
Because of thet we are not getting exception per each log event due to this issue, 
that is why it did not encounter during the test phases which we performed for the log appender.

By the way without allowing NPE we are throwing exception with the proper message.
Fixes are in https://github.com/wso2/shared-analytics/pull/126/files#diff-d45f010b040aded5b5711f7ddab28e59R438
 
Thanks
SajithD